groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -765,787 +1432,11454 @@ void Optional_DataImp<TYPE>::emplace(
+            BSLS_COMPILERFEATURES_FORWARD(ARGS_05, args_05),
+            BSLS_COMPILERFEATURES_FORWARD(ARGS_06, args_06),
+            BSLS_COMPILERFEATURES_FORWARD(ARGS_07, args_07),
+            BSLS_COMPILERFEATURES_FORWARD(ARGS_08, args_08),
+            BSLS_COMPILERFEATURES_FORWARD(ARGS_09, args_09));
+}
+#endif  // BSLSTL_OPTIONAL_VARIADIC_LIMIT_J >= 9
+
+#if BSLSTL_OPTIONAL_VARIADIC_LIMIT_J >= 10
+template <class TYPE, bool USES_BSLMA_ALLOC>
+template <class ARGS_01,
+          class ARGS_02,
+          class ARGS_03,
+          class ARGS_04,
+          class ARGS_05,
+          class ARGS_06,
+          class ARGS_07,
+          class ARGS_08,
+          class ARGS_09,
+          class ARGS_10>
+inline
+optional<TYPE, USES_BSLMA_ALLOC>::optional(
+                               bsl::in_place_t,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_01) args_01,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_02) args_02,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_03) args_03,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_04) args_04,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_05) args_05,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_06) args_06,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_07) args_07,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_08) args_08,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_09) args_09,
+                            BSLS_COMPILERFEATURES_FORWARD_REF(ARGS_10) args_10)
+{
+    emplace(BSLS_COMPILERFEATURES_FORWARD(ARGS_01, args_01),
+            BSLS_COMPILERFEATURES_FORWARD(ARGS_02, args_02),
+            BSLS_COMPILERFEATURES_FORWARD(ARGS_03, args_03),
+            BSLS_COMPILERFEATURES_FORWARD(ARGS_04, args_04),
+            BSLS_COMPILERFEATURES_FORWARD(ARGS_05, args_05),
+            BSLS_COMPILERFEATURES_FORWARD(ARGS_06, args_06),
+            BSLS_COMPILERFEATURES_FORWARD(ARGS_07, args_07),
+            BSLS_COMPILERFEATURES_FORWARD(ARGS_08, args_08),
+            BSLS_COMPILERFEATURES_FORWARD(ARGS_09, args_09),
+            BSLS_COMPILERFEATURES_FORWARD(ARGS_10, args_10));
+}
+#endif  // BSLSTL_OPTIONAL_VARIADIC_LIMIT_J >= 10
+
+
+#if defined(BSLS_COMPILERFEATURES_SUPPORT_GENERALIZED_INITIALIZERS)
+#if BSLSTL_OPTIONAL_VARIADIC_LIMIT_J >= 0
+template <class TYPE, bool USES_BSLMA_ALLOC>
+template <class INIT_LIST_TYPE>
+inline
+optional<TYPE, USES_BSLMA_ALLOC>::optional(
+                               bsl::in_place_t,
+                               std::initializer_list<INIT_LIST_TYPE>      il)
+{
+    emplace(il);
+}
+#endif  // BSLSTL_OPTIONAL_VARIADIC_LIMIT_J >= 0
+
+#if BSLSTL_OPTIONAL_VARIADIC_LIMIT_J >= 1
+template <class TYPE, bool USES_BSLMA_ALLOC>
+template <class INIT_LIST_TYPE, class ARGS_01>
+inline
+optional<TYPE, USES_BSLMA_ALLOC>::optional(
+       
```

Comment:
`std::optional` doesn't have get_allocator, right?  I think this will probably prompt some rewriting of this function and other swaps with std::optional below.

Reply:
std::optional doesn't, but bsl::optional does. The get_allocator method is only available in bsl::optional with AA ValueTypes (which matches std), and swap only works with same type, so we're good.

Reply:
There's some disconnect here.
In this code, `rhs` is a `std::optional<TYPE>&` and we do `rhs.get_allocator()`.  How is that supposed to compile if `std::optional<TYPE>` doesn't have a `get_allocator` method?

Reply:
good point, I'm assuming this crept in with the rework of swap. There shouldn't exist a swap function between a bsl::optional and an std::optional for an AA type. I'll remove the overloads for the AA types.

Reply:
I wonder if that's the thing that we should do, but maybe it is?  

I wonder if instead we could follow `NullableValue_(With/Without)Allocator::swap` approach to this:

1. Check if both optionals are empty - nothing to do.
2. Check if both are full - then use `bslalg::SwapUtil::swap` on the optional's values.  This will be efficient if the values share the (default) allocator and will do the copy-dance otherwise.
3. If only one of optionals has a value - move(?)-construct the one that is empty and reset one that is non-empty.

Thoughts?

Reply:
We could possibly do that for bsl::optional, if we choose to ignore that bsl::optional<AA_TYPE> is not the same type as std::optional<AA_TYPE>. The problem is that it may at some point be replaced with pmr::optional and I'm not entirely sure how LEWG will feel about this. I haven't dealt with them much, what is your hunch ?

My own analysis is :
- the problem in compatibility between `pmr::optional<AA_TYPE>` and `std::optional<AA_TYPE>` is that `pmr::optional` provides a guarantee that all the ValueType objects will be created with the designated allocator.  Swap between the two types which resorts to swapping the underlying values can't uphold that guarantee. However, swap when allocators differ is already undefined in the standard, so that's ok.
- the other issues is `swap(a,b)` vs `a.swap(b)` and b.swap(a)`. The standard requires free swap to be equivalent to invoking a member swap for conatiners. Optional is not a container. We do not have to provide it. If we do provide it, we can only do so for bsl::optional. I think it's ok to not provide it, and I think it's better that we don't at this point in time. We can provide it if the final pmr::optional provides one.  

TL;DR;
I think it's ok to do so, but I think we risk losing the functionality when pmr::optional comes out. 
Ideally, I would wait until pmr::optional has passed the LEWG stage and add it if LEWG is happy to do the same. 
WDYT ?

Reply:
I suggest we remove this functionality for now - we can add it back either after std::pmr::optional is done, or when we get requests for it.